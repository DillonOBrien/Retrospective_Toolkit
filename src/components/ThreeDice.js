import React, { Suspense } from 'react';
import { TextureLoader } from 'three';
import { Canvas, useLoader } from 'react-three-fiber';
import { GLTFLoader } from 'three/examples/jsm/loaders/GLTFLoader'
//import { SVGLoader as loader } from './SVGLoader';
import { Html, draco, OrbitControls } from 'drei';
import { Physics, useBox } from 'use-cannon';
import barChart from '../resources/bar_chart.png';
import bubbleChart from '../resources/bubble_chart.png';
import highlight from '../resources/highlight.png';
import insertEmoticon from '../resources/insert_emoticon.png';
import insertPhoto from '../resources/insert_photo.png';
import cloudQueue from '../resources/cloud_queue.png';

const Die = (props) => {
    const [mesh, api] = useBox(() => ({ mass: 1, position: [-6, 10, 3], velocity: [5, 0, -3] ,angularVelocity: [5, 0, 3], ...props}));
    return (
        <mesh ref={mesh}>
            <boxBufferGeometry attach='geometry' args={[1,1,1]} />
            <meshStandardMaterial attachArray='material' map={useLoader(TextureLoader, barChart)} />
            <meshStandardMaterial attachArray='material' map={useLoader(TextureLoader, bubbleChart)} />
            <meshStandardMaterial attachArray='material' map={useLoader(TextureLoader, highlight)} />
            <meshStandardMaterial attachArray='material' map={useLoader(TextureLoader, insertEmoticon)} />
            <meshStandardMaterial attachArray='material' map={useLoader(TextureLoader, insertPhoto)} />
            <meshStandardMaterial attachArray='material' map={useLoader(TextureLoader, cloudQueue)} />
        </mesh>
    )
}

const Plane = () => {
    return (
        <mesh>
            <planeBufferGeometry attach='geometry' args={[100,100]}/>
            <meshStandardMaterial attach='material' color='red'/>
        </mesh>
    )
}

function Model({ url, props }) {
    const [tray] = useBox(() => ({ 
        type: "Static",
        args: [12, 0, 22],
        position: [0, 1, 0], 
        rotation: [0, -Math.PI / 2, 0],
        ...props}));
    const { scene } = useLoader(GLTFLoader, url, draco())
    return (
        <primitive ref={tray} object={scene} dispose={null} />
    )
}

const Lights = () => {
    return (
        <>
            <ambientLight intensity={0.5} />
            <directionalLight
                intensity={1.0}
                position={[0, 20, 0]}
            />
        </>
    )
}

const ThreeDice = () => {

    return (
            <Canvas style={{width: '100vw', height: '500px'}} camera={{ position: [0, 20, 0], fov: 50 }}>             
                <Physics>
                    <Suspense fallback={<Html>loading..</Html>}>
                        <Die />
                        <Model url={'trayModel/tray.glb'} />
                    </Suspense>
                </Physics>
                <Lights />
                <gridHelper args={[50, 50, 'red', 'black']} />
                <OrbitControls />
            </Canvas>
    )
};

export default ThreeDice;