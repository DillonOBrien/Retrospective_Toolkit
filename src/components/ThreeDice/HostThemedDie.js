import React, { useEffect } from "react";
import { useTexture } from "@react-three/drei";
import { useBox } from "@react-three/cannon";
import PropTypes from "prop-types";
import { v4 as uuidv4 } from "uuid";
import { useGesture, useDrag } from "react-use-gesture";
// import { uniqueImageSet } from "../Dice/Dice";

const singleRollSound = () => {
  new Audio("/diceRoll.m4a").play();
};

// const newRandomArray = () => [Math.random(), Math.random(), Math.random()];

/* istanbul ignore next */
const ThemedDie = (props) => {
  const {
    diceInitPos,
    rerollAllToggle,
    rerollValue,
    rerollDieToggle,
    imageSet,
    // setImages,
    geom,
    setOrbitControl,
    mousePos,
    rotationValues,
    // setRotationValues,
  } = props;
  // eslint-disable-next-line no-console
  console.log(`Inside ThemedDie: ${imageSet}`);
  const textures = useTexture([...imageSet]);

  const [mesh, api] = useBox(() => ({
    mass: 300,
    inertia: 13,
    position: [diceInitPos[0], diceInitPos[1], diceInitPos[2]],
    rotation: [
      rotationValues[0] * Math.PI,
      rotationValues[1] * Math.PI,
      rotationValues[2] * Math.PI,
    ],
    linearDamping: 0.5,
    angularDamping: 0.1,
    material: { restitution: 0.3 },
  }));

  const isDragging = React.useRef(false);

  const clickBind = useGesture({
    onClick: () => {
      if (!isDragging.current) {
        rerollDieToggle(!rerollValue);
        singleRollSound();
      }
    },
  });

  const dragBind = useDrag(
    ({ first, last }) => {
      if (first) {
        setOrbitControl(false);
        isDragging.current = true;
        api.mass.set(0);
        api.collisionResponse.set(0);
      } else if (last) {
        // eslint-disable-next-line no-return-assign
        setTimeout(() => (isDragging.current = false), 100);
        setOrbitControl(true);
        api.mass.set(300);
        api.collisionResponse.set(1);
      }
      api.position.set(mousePos[0], 2, -mousePos[1]);
    },
    {
      initial: () => [mousePos[0], mousePos[1]],
      filterTaps: true,
    }
  );

  useEffect(() => {
    api.position.set(diceInitPos[0], diceInitPos[1], diceInitPos[2]);
    api.velocity.set(15, 0, -10);
    api.angularVelocity.set(-15, 2, -10);
    api.rotation.set(
      rotationValues[0] * Math.PI,
      rotationValues[1] * Math.PI,
      rotationValues[2] * Math.PI
    );
    // setRotationValues(newRandomArray);
    // setImages(uniqueImageSet);
  }, [
    api.angularVelocity,
    api.position,
    api.rotation,
    api.velocity,
    diceInitPos,
    rerollAllToggle,
    rerollValue,
    rotationValues,
  ]);

  return (
    <mesh
      ref={mesh}
      geometry={geom}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...dragBind()}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...clickBind()}
    >
      <boxBufferGeometry />
      {textures.map((image) => (
        <meshStandardMaterial
          key={uuidv4()}
          flatShading
          roughness={0.8}
          attachArray="material"
          map={image}
        />
      ))}
    </mesh>
  );
};

ThemedDie.propTypes = {
  diceInitPos: PropTypes.arrayOf(PropTypes.number).isRequired,
  imageSet: PropTypes.arrayOf(PropTypes.string).isRequired,
  setImages: PropTypes.func.isRequired,
  rerollAllToggle: PropTypes.bool.isRequired,
  rerollValue: PropTypes.bool.isRequired,
  rerollDieToggle: PropTypes.func.isRequired,
  geom: PropTypes.shape().isRequired,
  setOrbitControl: PropTypes.func.isRequired,
  mousePos: PropTypes.arrayOf(PropTypes.number).isRequired,
  rotationValues: PropTypes.arrayOf(PropTypes.number).isRequired,
  setRotationValues: PropTypes.func.isRequired,
};

export default ThemedDie;
