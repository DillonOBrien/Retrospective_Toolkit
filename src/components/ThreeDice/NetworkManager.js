/* eslint-disable no-console */
import React, { useState, useEffect } from "react";
import { useParams } from "react-router-dom";
import * as io from "socket.io-client";
import DiceLanding from "../../containers/DiceLanding";

const NetworkManager = () => {
  const [socket, setSocket] = useState();
  const { roomId } = useParams();
  console.log("rendering");

  // check with server to see if roomId exists
  // if () === null) {};
  // redirect to page not found

  useEffect(() => {
    setSocket(io(window.location.origin.replace(window.location.port, "5000")));
  }, []);

  // eslint-disable-next-line consistent-return
  useEffect(() => {
    if (socket) {
      console.log("on fired");
      socket.emit("board:create", roomId);
      console.log("emit fired");
      socket.on("create:complete", (msg) => {
        console.log(msg);
      });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [socket]);

  return <>{socket && <DiceLanding socket={socket} roomId={roomId} />}</>;
};

export default NetworkManager;
