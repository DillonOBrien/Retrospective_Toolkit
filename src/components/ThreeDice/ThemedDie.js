/* eslint-disable no-console */
import React, { useEffect } from "react";
import { useThree } from "react-three-fiber";
import { useTexture } from "@react-three/drei";
import { useBox } from "@react-three/cannon";
import PropTypes from "prop-types";
import { v4 as uuidv4 } from "uuid";
import { useGesture, useDrag } from "react-use-gesture";
import { uniqueImageSet } from "../Dice/Dice";

const singleRollSound = () => {
  new Audio("/diceRoll.m4a").play();
};

/* istanbul ignore next */
const ThemedDie = (props) => {
  const {
    diceInitPos,
    rerollAllToggle,
    rerollValue,
    rerollDieToggle,
    imageSet,
    setImages,
    geom,
    setOrbitControl,
  } = props;
  // const [pos, setPos] = useState(diePos);
  const textures = useTexture([...imageSet]);

  const [mesh, api] = useBox(() => ({
    mass: 300,
    inertia: 13,
    rotation: [
      Math.random() * Math.PI,
      Math.random() * Math.PI,
      Math.random() * Math.PI,
    ],
    linearDamping: 0.5,
    angularDamping: 0.1,
    material: { restitution: 0.3 },
  }));

  const { size, viewport } = useThree();
  const aspect = size.width / viewport.width;
  const isDragging = React.useRef(false);

  const clickBind = useGesture({
    onClick: () => {
      if (isDragging.current) return;
      rerollDieToggle(!rerollValue);
      singleRollSound();
    },
  });

  const dragBind = useDrag(
    ({ offset: [x, y], first, last }) => {
      if (first) {
        setOrbitControl(false);
        isDragging.current = true;
      } else if (last) {
        isDragging.current = false;
        setOrbitControl(true);
      }
      api.position.set(x / aspect, 1.5, y / aspect);
    },
    {
      threshold: 1,
      delay: 500,
    }
  );

  useEffect(() => {
    api.position.set(diceInitPos[0], diceInitPos[1], diceInitPos[2]);
    api.velocity.set(15, 0, -10);
    api.angularVelocity.set(-15, 2, -10);
    setImages(uniqueImageSet);
  }, [
    api.angularVelocity,
    api.position,
    api.velocity,
    diceInitPos,
    rerollAllToggle,
    rerollValue,
    setImages,
  ]);

  return (
    <mesh
      ref={mesh}
      geometry={geom}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...dragBind()}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...clickBind()}
    >
      <boxBufferGeometry />
      {textures.map((image) => (
        <meshStandardMaterial
          key={uuidv4()}
          flatShading
          roughness={0.8}
          attachArray="material"
          map={image}
        />
      ))}
    </mesh>
  );
};

ThemedDie.propTypes = {
  diceInitPos: PropTypes.arrayOf(PropTypes.number).isRequired,
  imageSet: PropTypes.arrayOf(PropTypes.string).isRequired,
  setImages: PropTypes.func.isRequired,
  rerollAllToggle: PropTypes.bool.isRequired,
  rerollValue: PropTypes.bool.isRequired,
  rerollDieToggle: PropTypes.func.isRequired,
  geom: PropTypes.shape().isRequired,
  setOrbitControl: PropTypes.func.isRequired,
};

export default ThemedDie;
