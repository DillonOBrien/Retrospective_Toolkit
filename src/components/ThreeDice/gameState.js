import { atom } from "jotai";
import { atomWithReset } from "jotai/utils";
import { v4 as uuidv4 } from "uuid";

export const gameStart = atom(false);
export const gameReset = atom(false);

export const diceDefault = atom([
  {
    uuid: uuidv4(),
    position: [-13, 5, 6],
  },
  {
    uuid: uuidv4(),
    position: [-13, 6, 4],
  },
  {
    uuid: uuidv4(),
    position: [-13, 7, 5],
  },
  {
    uuid: uuidv4(),
    position: [-13, 5, 5],
  },
  {
    uuid: uuidv4(),
    position: [-13, 6, 2],
  },
]);

export const diceArray = atomWithReset([[]]);
