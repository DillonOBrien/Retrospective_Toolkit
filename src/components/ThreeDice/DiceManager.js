import React from "react";
import { useAtom } from "jotai";
import { atomWithReset } from "jotai/utils";
import { v4 as uuidv4 } from "uuid";
import { uniqueImageSet } from "../Dice/Dice";
import ThemedDie from "./ThemedDie";
import { diceDefaultState, rerollState } from "./gameState";

const DiceManager = () => {
  const diceImageState = atomWithReset([
    {
      uuid: [uuidv4(), uuidv4(), uuidv4(), uuidv4(), uuidv4(), uuidv4()],
      url: uniqueImageSet(),
    },
    {
      uuid: [uuidv4(), uuidv4(), uuidv4(), uuidv4(), uuidv4(), uuidv4()],
      url: uniqueImageSet(),
    },
    {
      uuid: [uuidv4(), uuidv4(), uuidv4(), uuidv4(), uuidv4(), uuidv4()],
      url: uniqueImageSet(),
    },
    {
      uuid: [uuidv4(), uuidv4(), uuidv4(), uuidv4(), uuidv4(), uuidv4()],
      url: uniqueImageSet(),
    },
    {
      uuid: [uuidv4(), uuidv4(), uuidv4(), uuidv4(), uuidv4(), uuidv4()],
      url: uniqueImageSet(),
    },
  ]);

  const [dicePosition] = useAtom(diceDefaultState);
  const [images] = useAtom(diceImageState);
  const [reroll] = useAtom(rerollState);

  return dicePosition.map((pos, index) => (
    <ThemedDie
      key={pos.uuid}
      theme="random"
      dicePos={pos.position}
      rerollToggle={reroll}
      dieIndex={index}
      imageSet={images}
    />
  ));
};

export default DiceManager;
