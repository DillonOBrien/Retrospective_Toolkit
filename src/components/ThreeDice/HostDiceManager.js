/* eslint-disable no-console */
import React, { Suspense, useEffect, useMemo, useState } from "react";
import { useAtom } from "jotai";
import PropTypes from "prop-types";
import { BoxBufferGeometry } from "three";
import { useThree, useFrame } from "react-three-fiber";
import { uniqueImageSet } from "../Dice/Dice";
import HostThemedDie from "./HostThemedDie";
import CollisionMesh from "./CollisionMesh";
import { diceDefaultState } from "./gameState";

/* istanbul ignore next */
const DiceManager = (props) => {
  const { reroll, setOrbitControl, socket, roomId } = props;
  const geom = useMemo(() => new BoxBufferGeometry(), []);

  const [dicePosition] = useAtom(diceDefaultState);

  const [userGameReady, setUserReady] = useState(false);

  const [dieImagesOne, setImagesOne] = useState(uniqueImageSet);
  const [dieImagesTwo, setImagesTwo] = useState(uniqueImageSet);
  const [dieImagesThree, setImagesThree] = useState(uniqueImageSet);
  const [dieImagesFour, setImagesFour] = useState(uniqueImageSet);
  const [dieImagesFive, setImagesFive] = useState(uniqueImageSet);

  const [rerollOne, toggleRerollOne] = useState(false);
  const [rerollTwo, toggleRerollTwo] = useState(false);
  const [rerollThree, toggleRerollThree] = useState(false);
  const [rerollFour, toggleRerollFour] = useState(false);
  const [rerollFive, toggleRerollFive] = useState(false);

  const [rotationOne, setRotationOne] = useState([
    Math.random(),
    Math.random(),
    Math.random(),
  ]);
  const [rotationTwo, setRotationTwo] = useState([
    Math.random(),
    Math.random(),
    Math.random(),
  ]);
  const [rotationThree, setRotationThree] = useState([
    Math.random(),
    Math.random(),
    Math.random(),
  ]);
  const [rotationFour, setRotationFour] = useState([
    Math.random(),
    Math.random(),
    Math.random(),
  ]);
  const [rotationFive, setRotationFive] = useState([
    Math.random(),
    Math.random(),
    Math.random(),
  ]);

  const imageArray = [
    { images: dieImagesOne, setImages: setImagesOne },
    { images: dieImagesTwo, setImages: setImagesTwo },
    { images: dieImagesThree, setImages: setImagesThree },
    { images: dieImagesFour, setImages: setImagesFour },
    { images: dieImagesFive, setImages: setImagesFive },
  ];

  const userImageArray = [
    dieImagesOne,
    dieImagesTwo,
    dieImagesThree,
    dieImagesFour,
    dieImagesFive,
  ];

  const userRotationArray = [
    rotationOne,
    rotationTwo,
    rotationThree,
    rotationFour,
    rotationFive,
  ];

  const rerollArray = [
    { rerollDie: rerollOne, reroll: toggleRerollOne },
    { rerollDie: rerollTwo, reroll: toggleRerollTwo },
    { rerollDie: rerollThree, reroll: toggleRerollThree },
    { rerollDie: rerollFour, reroll: toggleRerollFour },
    { rerollDie: rerollFive, reroll: toggleRerollFive },
  ];

  const rotationArray = [
    { rotation: rotationOne, setRotation: setRotationOne },
    { rotation: rotationTwo, setRotation: setRotationTwo },
    { rotation: rotationThree, setRotation: setRotationThree },
    { rotation: rotationFour, setRotation: setRotationFour },
    { rotation: rotationFive, setRotation: setRotationFive },
  ];

  useEffect(() => {
    console.log(`Inside useEffect: ${imageArray[0].images}`);
    if (userGameReady)
      socket.emit("host:newRoll", roomId, userRotationArray, userImageArray);
    else {
      socket.emit("game:start", roomId, userRotationArray, userImageArray);
      setUserReady(true);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const { viewport } = useThree();
  const mousePos = [];

  useFrame((state) => {
    const { mouse } = state;
    const { width, height } = viewport();
    mousePos[0] = (mouse.x * width) / 2;
    mousePos[1] = (mouse.y * height) / 2;
  });

  return (
    <Suspense fallback={null}>
      {dicePosition.map((pos, index) => (
        <HostThemedDie
          key={pos.uuid}
          diceInitPos={pos.position}
          rerollAllToggle={reroll}
          rerollValue={rerollArray[index].rerollDie}
          rerollDieToggle={rerollArray[index].reroll}
          imageSet={imageArray[index].images}
          setImages={imageArray[index].setImages}
          geom={geom}
          setOrbitControl={setOrbitControl}
          mousePos={mousePos}
          rotationValues={rotationArray[index].rotation}
          setRotationValues={rotationArray[index].setRotation}
        />
      ))}
      <CollisionMesh />
    </Suspense>
  );
};

DiceManager.propTypes = {
  reroll: PropTypes.bool,
  setOrbitControl: PropTypes.func.isRequired,
  roomId: PropTypes.string.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  socket: PropTypes.object.isRequired,
};

DiceManager.defaultProps = {
  reroll: false,
};

export default DiceManager;
