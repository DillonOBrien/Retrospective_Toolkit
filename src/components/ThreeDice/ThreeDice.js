import React, { Suspense, useEffect } from "react";
import { Canvas } from "react-three-fiber";
import { Html, useGLTF, useTexture, OrbitControls, useProgress } from "drei";
// eslint-disable-next-line import/no-unresolved
import { Physics, useBox } from "@react-three/cannon";
import Button from "@material-ui/core/Button";
import { makeStyles } from "@material-ui/core/styles";
import Icon from "@material-ui/core/Icon";
import { Provider, useAtom } from "jotai";
import PropTypes from "prop-types";
import { gameStartState, diceDefaultState, rerollState } from "./gameState";
import { randomDiceThemes as themes } from "../../constants/DieConstants";

const CollisionMesh = () => {
  const [floor] = useBox(() => ({
    position: [0.5, 0, 0],
    rotation: [0, -Math.PI / 2, 0],
    args: [14, 0, 22],
    material: { friction: 1000 },
  }));
  const [left] = useBox(() => ({
    position: [-10.5, 1, 0],
    rotation: [0, -Math.PI / 2, 0],
    args: [13, 4, 0],
  }));
  const [right] = useBox(() => ({
    position: [11.3, 1, 0],
    rotation: [0, -Math.PI / 2, 0],
    args: [13, 4, 0],
  }));
  const [top] = useBox(() => ({
    position: [0.5, 1, 6.5],
    args: [22, 4, 0],
  }));
  const [bottom] = useBox(() => ({
    position: [0.5, 1, -6.5],
    args: [22, 4, 0],
  }));
  return (
    <group>
      <mesh ref={floor}>
        <boxBufferGeometry args={[14, 0, 22]} />
        <meshStandardMaterial color="red" transparent opacity={0} />
      </mesh>
      <mesh ref={left}>
        <boxBufferGeometry args={[13, 4, 0]} />
        <meshStandardMaterial color="red" transparent opacity={0} />
      </mesh>
      <mesh ref={right}>
        <boxBufferGeometry args={[13, 4, 0]} />
        <meshStandardMaterial color="red" transparent opacity={0} />
      </mesh>
      <mesh ref={top}>
        <boxBufferGeometry args={[22, 4, 0]} />
        <meshStandardMaterial color="red" transparent opacity={0} />
      </mesh>
      <mesh ref={bottom}>
        <boxBufferGeometry args={[22, 4, 0]} />
        <meshStandardMaterial color="red" transparent opacity={0} />
      </mesh>
    </group>
  );
};

const Loader = () => {
  const { progress } = useProgress();
  return <Html center>{Math.trunc(progress)} % loaded</Html>;
};

const ThemedDie = (props) => {
  const { theme, dicePos, rerollToggle } = props;
  const actionTextures = useTexture([...themes.URL]);
  // eslint-disable-next-line no-console
  console.log("Action Textures");
  // eslint-disable-next-line no-console
  console.log(actionTextures);

  const [mesh, api] = useBox(() => ({
    mass: 300,
    inertia: 13,
    rotation: [
      Math.random() * Math.PI,
      Math.random() * Math.PI,
      Math.random() * Math.PI,
    ],
    linearDamping: 0.5,
    angularDamping: 0.1,
    material: { restitution: 0.3 },
  }));

  useEffect(() => {
    api.position.set(dicePos[0], dicePos[1], dicePos[2]);
    api.velocity.set(15, 0, -10);
    api.angularVelocity.set(-15, 2, -10);
  }, [api.angularVelocity, api.position, api.velocity, dicePos, rerollToggle]);

  if (theme === "action") {
    return (
      <mesh
        onClick={() => {
          api.position.set(dicePos[0], dicePos[1], dicePos[2]);
          api.velocity.set(15, 0, -10);
          api.angularVelocity.set(-15, 2, -10);
        }}
        ref={mesh}
      >
        <boxBufferGeometry />
        {actionTextures.map((image) => (
          <meshStandardMaterial
            key={image.uuid}
            flatShading
            roughness={0.8}
            attachArray="material"
            map={image}
          />
        ))}
      </mesh>
    );
  }
  // if we are here, something has gone wrong
  return new Error("problem encountered in ThemedDice");
};
ThemedDie.propTypes = {
  theme: PropTypes.string.isRequired,
  dicePos: PropTypes.arrayOf(PropTypes.number).isRequired,
  rerollToggle: PropTypes.bool.isRequired,
};

const GameManager = () => {
  const [gameStarted, setGameState] = useAtom(gameStartState);
  const [reroll, rerollDice] = useAtom(rerollState);
  const [dicePosition] = useAtom(diceDefaultState);

  const useStyles = makeStyles((theme) => ({
    button: {
      margin: theme.spacing(1),
      width: "120px",
    },
    buttonStart: {
      margin: theme.spacing(1),
      width: "160px",
    },
  }));

  const classes = useStyles();

  return (
    <>
      <ambientLight intensity={0.5} />
      <directionalLight
        intensity={1.0}
        position={[0, 20, 0]}
        shadow-mapSize-width={1024}
        shadow-mapSize-height={1024}
      />
      <Suspense fallback={<Loader />}>
        <Model url="trayModel/tray.glb" />
      </Suspense>
      {!gameStarted && (
        <Html position={[-4, 0, 2]} scaleFactor={25}>
          <Button
            variant="contained"
            color="primary"
            className={classes.buttonStart}
            endIcon={<Icon>casino</Icon>}
            onClick={() => {
              setGameState(true);
            }}
          >
            Start Game
          </Button>
        </Html>
      )}
      {gameStarted && (
        <>
          <Suspense fallback={null}>
            {dicePosition.map((pos) => (
              <ThemedDie
                key={pos.uuid}
                theme="action"
                dicePos={pos.position}
                rerollToggle={reroll}
              />
            ))}
            <CollisionMesh />
          </Suspense>
          <Html position={[-3, 0, 9]} scaleFactor={25}>
            <Button
              variant="contained"
              color="primary"
              className={classes.button}
              endIcon={<Icon>casino</Icon>}
              onClick={() => {
                rerollDice(!reroll);
              }}
            >
              Roll It!
            </Button>
          </Html>
        </>
      )}
    </>
  );
};

function Model({ url }) {
  const { scene } = useGLTF(url);
  return (
    <primitive rotation={[0, -Math.PI / 2, 0]} object={scene} dispose={null} />
  );
}

Model.propTypes = {
  url: PropTypes.string.isRequired,
};

const ThreeDice = () => (
  <Canvas
    concurrent
    style={{ width: "100vw", height: "500px" }}
    camera={{ position: [0, 20, 12], fov: 50 }}
  >
    <Provider>
      <Physics gravity={[0, -30, 0]} defaultContactMaterial>
        <GameManager />
      </Physics>
    </Provider>
    <OrbitControls />
  </Canvas>
);

export default ThreeDice;
