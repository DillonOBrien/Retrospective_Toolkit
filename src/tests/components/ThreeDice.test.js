/* eslint no-unused-vars: 0 */
import React from "react";
import renderer from "react-test-renderer";
import { install } from "resize-observer";
import { shallow } from "enzyme";
import Button from "@material-ui/core/Button";
import {
  render,
  getByText,
  getByLabelText,
  screen,
} from "@testing-library/react";
import ThreeDice from "../../components/ThreeDice/ThreeDice";

beforeAll(() => {
  if (!window.ResizeObserver) install();
  jest.spyOn(console, "warn").mockImplementation(() => {});
  jest.spyOn(console, "error").mockImplementation(() => {});
  // jest.spyOn(console, "log").mockImplementation(() => {});
});

test("SNAPSHOT TEST THREEDICE", () => {
  const testRenderer = renderer.create(<ThreeDice />);

  expect(testRenderer.toTree()).toMatchSnapshot();
});

test("Custom Button and Loading Html elements are loading", () => {
  expect(screen.getByText("roll")).toBe(true);
  expect(getByText.contains("Button")).toBe(true);
  expect(getByText.contains("Html")).toBe(true);
});
