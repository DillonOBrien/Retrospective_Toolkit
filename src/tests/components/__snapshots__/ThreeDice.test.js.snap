// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SNAPSHOT TEST THREEDICE 1`] = `
Object {
  "instance": null,
  "nodeType": "component",
  "props": Object {},
  "rendered": Object {
    "instance": null,
    "nodeType": "component",
    "props": Object {
      "camera": Object {
        "fov": 30,
        "position": Array [
          0,
          20,
          15,
        ],
      },
      "children": Array [
        <Provider>
          <UNDEFINED
            defaultContactMaterial={true}
            gravity={
              Array [
                0,
                -30,
                0,
              ]
            }
          >
            <GameManager />
          </UNDEFINED>
        </Provider>,
        <ForwardRef />,
      ],
      "concurrent": true,
      "style": Object {
        "height": "500px",
        "width": "100vw",
      },
    },
    "rendered": Object {
      "instance": null,
      "nodeType": "component",
      "props": Object {
        "camera": Object {
          "fov": 30,
          "position": Array [
            0,
            20,
            15,
          ],
        },
        "children": Array [
          <Provider>
            <UNDEFINED
              defaultContactMaterial={true}
              gravity={
                Array [
                  0,
                  -30,
                  0,
                ]
              }
            >
              <GameManager />
            </UNDEFINED>
          </Provider>,
          <ForwardRef />,
        ],
        "concurrent": true,
        "preRender": <canvas
          style={
            Object {
              "display": "block",
            }
          }
        />,
        "renderer": [Function],
        "style": Object {
          "height": "500px",
          "width": "100vw",
        },
      },
      "rendered": Object {
        "instance": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <canvas
              style={
                Object {
                  "display": "block",
                }
              }
            />,
            false,
          ],
          "style": Object {
            "height": "500px",
            "overflow": "hidden",
            "position": "relative",
            "width": "100vw",
          },
        },
        "rendered": Array [
          Object {
            "instance": null,
            "nodeType": "host",
            "props": Object {
              "style": Object {
                "display": "block",
              },
            },
            "rendered": Array [],
            "type": "canvas",
          },
        ],
        "type": "div",
      },
      "type": [Function],
    },
    "type": [Function],
  },
  "type": [Function],
}
`;
